python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,395 @@
+                return self.owner_.indx_
+
+            def get_dependent_str(self):
+                name = ""
+                for dependent in self.dependent_:
+                    mname, dname = dependent.get_name()
+                    name = name + (mname + ":output(" + dname if self.itype_ == 2 else "")
+                    name = name + (")" if self.itype_ == 2 else mname + ":" + dname)
+                return name
+
+            def add_dependent(self, dependent):
+                """
+                # check if the dependency setting correct.
+                # correct connection are following
+                # 1. global input to module input
+                # 2. module output to next module input
+                # 3. module output to global output
+                """
+                owner_indx = self.get_owner_indx()
+                dep_owner_indx = dependent.get_owner_indx()
+                assert owner_indx != dep_owner_indx, f"can not set self as dependent."
+                assert not (
+                    owner_indx > dep_owner_indx
+                    and not (dependent.itype_ == 2 and dep_owner_indx == 0)
+                ), f"dependent only can be next module interface or global output."
+                assert not (
+                    owner_indx == 0 and dependent.itype_ != 1
+                ), f"global input only can set dependent with module input."
+
+                self.dependent_.append(dependent)
+
+        def __init__(self, indx=0):
+            self.indx_ = indx
+            self.name_ = "mod" + str(indx) if indx else ""
+            self.interfaces_ = {1: {}, 2: {}}
+            self.target_host_ = None
+            self.mod_name_ = "default"
+            self.build_func_ = None
+            self.params_ = None
+            self.target_ = None
+            self.dev_ = None
+
+
+        def get_interface(self, itype, name):
+            if name not in self.interfaces_[itype]:
+                self.interfaces_[itype][name] = self.Interface(self, itype, name)
+
+            return self.interfaces_[itype][name]
+
+        def input(self, name):
+            return self.get_interface(1, name)
+
+        def output(self, index):
+            return self.get_interface(2, index)
+
+        def set_target_host(self, host):
+            self.target_host_ = host
+
+        def set_mod_name(self, name):
+            self.mod_name_ = name
+
+        def set_build_func(self, build_func):
+            self.build_func_ = build_func
+
+        def set_params(self, params):
+            self.params_ = params
+
+        def set_target(self, target):
+            self.target_ = target
+
+        def set_dev(self, dev):
+            self.dev_ = dev
+
+    def __init__(self, mods):
+        self.pipe_module_name_ = "pipeline_module"
+        self.mod_wrapper = {m: self.ModuleWrapper(i + 1) for m, i in zip(mods, range(len(mods)))}
+        self.mod_wrapper[self.pipe_module_name_] = self.ModuleWrapper(0)
```

Comment:
As I mentioned before, I don't think we need to enforce the pipeline inputs to be the inputs of the first module. For example, we should support:

```
data0 -> mod0 -> mod1 -> output
data1 -------------|
```

Reply:
self.ModuleWrapper(0) is not for first sub module, this object is the global module that represent pipelinemodule and use for exposing global input and output interface/interface binding, 

the idea here is to invent a virtual module beside of all sub module to storage the connection relations between pipeline executor input/output and internal sub module  for example 
```
    input:
         data_0 : mod0: data_0
    output:
         ouput(0) : mod0:ouput(0) 
```

Reply:
That makes sense, but this implementation looks confusing, especially module index plays an important role in the current configuration. Since users are not supposed to access the global module, maybe we could just put it separately, like

```
self.root_mod_wrapper = self.ModuleWrapper()
self.mod_wrappers = {}
```

So that we could also make sure `root_mod_wrapper` will not be accessed by users via provided APIs such as `__getitem__`. This also addresses the inconsistent type of the current `self.mod_wrapper: Dict[Union[str, Module], ModuleWrapper]`.

Reply:
yes, this more clear, I would apply the change.